images/tools/database-copy/entrypoint.sh

```diff
@@ -1,30 +1,102 @@
+    exit $exit_code
+  fi
 
-while [ $CHECK_COUNT -lt $CHECK_NUMBER ]; do
-  sleep $CHECK_INTERVAL
-  TASKS_RUNNING="$(ps -e -o pid,comm | grep -c "$CLIENT_TASK")"
+  aws s3 cp data_dump.sql s3://${S3_BUCKET_NAME}/
+  exit_code=$?
 
-  if [[ $TASKS_RUNNING == 0 ]]; then
-     CHECK_COUNT=$(( $CHECK_COUNT + 1 ))
-     TIME_TO_SHUTDOWN="$(( (CHECK_NUMBER - CHECK_COUNT) * CHECK_INTERVAL ))"
-     echo "No clients connected, will shutdown in approximately $TIME_TO_SHUTDOWN seconds"
-  else
-     CHECK_COUNT=0
-     echo "$TASKS_RUNNING clients are connected"
+  if [ ${exit_code} -ne 0 ]
+  then
+    echo "Aborting data dump"
+    exit $exit_code
   fi
-done
 
-# Trigger CloudFormation stack delete before shutting down
-if [[ ! -z $ECS_CONTAINER_METADATA_URI_V4 ]]; then
-  aws cloudformation delete-stack --stack-name task-$(curl $ECS_CONTAINER_METADATA_URI_V4 -s | jq -r ".Name")
-fi
+  echo "Stopping data dump"
+else
+  echo "Starting data load"
+
+  echo "Copying data dump from S3"
+  aws s3 cp s3://${S3_BUCKET_NAME}/data_dump.sql data_dump.sql
+  
+  handle_errors $? "Copy failed"
+
+  echo "Scaling down services"
+  SERVICES_DATA=$(aws ecs list-services --cluster "${ECS_CLUSTER}")
+  handle_errors $? "Failed to list services"
+  SERVICES=$(echo "${SERVICES_DATA}" | jq -r '.serviceArns[]')
 
-echo "Shutting down"
+  for service in ${SERVICES}
+  do
+    COUNT_DATA=$(aws ecs describe-services --cluster "${ECS_CLUSTER}" --services "${service}")
+    handle_errors $? "Failed to describe service"
+    COUNT=$(echo "${COUNT_DATA}" | jq '.services[0].desiredCount')
+
+    SERVICE_NAME=$(basename "${service}")
+    CONFIG_FILE="${SERVICE_NAME}.desired_count"
+    echo "${COUNT}" > "${CONFIG_FILE}"
+
+    echo ${SERVICE_NAME}
+    UPDATE_DATA=$(aws ecs update-service --cluster "${ECS_CLUSTER}" --service "${service}" --desired-count 0)
+    handle_errors $? "Failed to update service ${SERVICE_NAME}"
+    echo "${UPDATE_DATA}" | jq -r '"  Desired Count: \(.service.desiredCount)\n  Running Count: \(.service.runningCount)"'
+  
+  done
+
+  echo "Clearing down the database prior to loading new data"
+  psql "${DB_CONNECTION_STRING}" -f /clear_db.sql
+
+  handle_errors $? "Clear down failed"
+  
+  echo "Restoring data from dump file"
+  pg_restore --format c --dbname "${DB_CONNECTION_STRING}" data_dump.sql
+  
+  handle_errors $? "Restore failed"
+  for service in ${SERVICES}
+  do
+    CONFIG_FILE="$(basename "${service}").desired_count"
+    COUNT=$(cat "${CONFIG_FILE}")
+    SERVICE_NAME=$(basename "${service}")
+    echo "Scaling up services"
+    echo ${SERVICE_NAME}
+    UPDATE_DATA=$(aws ecs update-service --cluster "${ECS_CLUSTER}" --service "${service}" --desired-count "${COUNT}")
+    handle_errors $? "Failed to update service ${SERVICE_NAME}"
+    echo "${UPDATE_DATA}" | jq -r '"  Desired Count: \(.service.desiredCount)\n  Running Count: \(.service.runningCount)"'
+  done
+
+  clean_up
+
+  echo "Stopping data load"
+fi
```

Comment:
This bash script is quite unwieldy.  Could we separate it into scripts for the different 'steps'?  Maybe should consider something like https://aws.amazon.com/step-functions/features/?pg=ln&sec=hs in the future.

Reply:
We'll look at this in the tidy up step.